# Нормализация признаков

###### Импорт библиотек

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

###### Импорт датасета и разбиение на X_train, X_test и y_train, y_test

In [34]:
df_train = pd.read_csv('perceptron-train.csv', header=None)
df_test = pd.read_csv('perceptron-test.csv', header=None)
df_train # - в первом столбце - класс, во втором и третьем - признаки

0         1            2
0   -1.0 -0.024626  1174.600238
1    1.0 -0.978058  1083.198803
2   -1.0  0.314272 -1472.977609
3   -1.0  0.179752   231.017267
4    1.0 -1.262544  -778.271726
..   ...       ...          ...
295 -1.0 -0.126438  1861.564124
296 -1.0 -0.490972  1815.652064
297  1.0 -0.375750  -314.074374
298  1.0 -1.861976 -1083.677679
299  1.0 -0.039329 -1863.884299

[300 rows x 3 columns]

In [41]:
X_train, y_train = df_train[[1, 2]], df_train[0].astype(int)
X_test, y_test = df_test[[1, 2]], df_test[0].astype(int)

###### Применение классификации (персептрона) к ненормализованной выборке

In [53]:
# создание модели с параметрами по умолчанию
perceptron_nonorm = Perceptron(random_state=241, max_iter=5, tol=None)
perceptron_nonorm.fit(X_train, y_train)

Perceptron(max_iter=5, random_state=241, tol=None)

In [54]:
# оценка качества модели с применением метрики "точность"
accuracy_nonorm = accuracy_score(y_test, perceptron_nonorm.predict(X_test))
print('Качество модели без нормализации признаков -', round(accuracy_nonorm, 2))

Качество модели без нормализации признаков - 0.66


###### Применение классификации (персептрона) к нормализованной выборке

In [55]:
# нормализация признаков с применением стандартизации
Scaler = StandardScaler()
X_train_norm = Scaler.fit_transform(X_train)
X_test_norm = Scaler.transform(X_test)

In [56]:
# создание модели с параметрами по умолчанию
perceptron_norm = Perceptron(random_state=241, max_iter=5, tol=None)
perceptron_norm.fit(X_train_norm, y_train)

Perceptron(max_iter=5, random_state=241, tol=None)

In [57]:
# оценка качества модели с нормализованными признаками с применением метрики "точность"
accuracy_norm = accuracy_score(y_test, perceptron_norm.predict(X_test_norm))
print('Качество модели с нормализованными признаками -', round(accuracy_norm, 2))

Качество модели с нормализованными признаками - 0.84


In [65]:
print('Разница качества - ', round(accuracy_norm % accuracy_nonorm * 100), '%', sep='')

Разница качества - 18%


Нормализация признаков дает существенное увеличение качества модели: точность увеличилась с $0.66$ до $0.84$ или на $18$$\%$.